File: src/main/java/io/github/tamawish/pureeconomy/util/Schedulers.java
package io.github.tamawish.pureeconomy.util;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitTask;

/**
 * Folia-aware scheduling helpers.
 *
 * <p>Uses Paper {@code GlobalRegionScheduler} / {@code AsyncScheduler} / entity schedulers when
 * present, and falls back to the Bukkit scheduler on Spigot.
 */
public final class Schedulers {

  private static final boolean FOLIA_LIKE = hasGlobal();

  private Schedulers() {}

  private static boolean hasGlobal() {
    try {
      Bukkit.class.getMethod("getGlobalRegionScheduler");
      return true;
    } catch (NoSuchMethodException e) {
      return false;
    }
  }

  /**
   * Runs a task on the global region (Paper/Folia) or the main server thread (Spigot).
   *
   * @param plugin owning plugin
   * @param task work to execute
   */
  public static void runGlobal(Plugin plugin, Runnable task) {
    if (FOLIA_LIKE) {
      Bukkit.getGlobalRegionScheduler().execute(plugin, task);
    } else {
      Bukkit.getScheduler().runTask(plugin, task);
    }
  }

  /**
   * Schedules a repeating global-region (or main-thread) timer.
   *
   * @param plugin owning plugin
   * @param task work to execute each period
   * @param delayTicks ticks before the first run
   * @param periodTicks ticks between runs
   * @return opaque task handle suitable for {@link #cancel(Object)}
   */
  public static Object runGlobalTimer(
      Plugin plugin, Runnable task, long delayTicks, long periodTicks) {
    if (FOLIA_LIKE) {
      return Bukkit.getGlobalRegionScheduler()
          .runAtFixedRate(
              plugin, scheduled -> task.run(), Math.max(1L, delayTicks), Math.max(1L, periodTicks));
    }
    return Bukkit.getScheduler().runTaskTimer(plugin, task, delayTicks, periodTicks);
  }

  /**
   * Runs a task asynchronously for I/O or other off-thread work.
   *
   * @param plugin owning plugin
   * @param task work to execute off the server thread
   */
  public static void runAsync(Plugin plugin, Runnable task) {
    try {
      Bukkit.getAsyncScheduler().runNow(plugin, scheduled -> task.run());
    } catch (NoSuchMethodError | NoClassDefFoundError e) {
      Bukkit.getScheduler().runTaskAsynchronously(plugin, task);
    }
  }

  /**
   * Runs a delayed task on the player's owning region (Folia) or the main thread (Spigot).
   *
   * @param plugin owning plugin
   * @param player entity whose region owns the task
   * @param task work to execute
   * @param delayTicks ticks to wait before running
   */
  public static void runAtEntityLater(
      Plugin plugin, Player player, Runnable task, long delayTicks) {
    try {
      player
          .getScheduler()
          .runDelayed(plugin, scheduled -> task.run(), null, Math.max(1L, delayTicks));
    } catch (NoSuchMethodError | NoClassDefFoundError e) {
      Bukkit.getScheduler().runTaskLater(plugin, task, Math.max(1L, delayTicks));
    }
  }

  /**
   * Cancels a task returned by {@link #runGlobalTimer(Plugin, Runnable, long, long)}.
   *
   * @param task opaque Folia or Bukkit task handle; ignored when {@code null}
   */
  public static void cancel(Object task) {
    if (task == null) {
      return;
    }
    try {
      task.getClass().getMethod("cancel").invoke(task);
      return;
    } catch (Exception ignored) {
      // Fall through to BukkitTask handling.
    }
    if (task instanceof BukkitTask bukkitTask) {
      bukkitTask.cancel();
    }
  }
}

File: src/main/kotlin/io/github/tamawish/pureeconomy/util/Schedulers.kt
package io.github.tamawish.pureeconomy.util

import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitTask

/**
 * Folia-aware scheduling helpers.
 *
 * When Paper/Folia region schedulers are present, every path uses them and never calls
 * [org.bukkit.scheduler.BukkitScheduler] (which throws on Folia). Spigot/Bukkit fall back to the
 * classic scheduler.
 */
object Schedulers {
    private val FOLIA_LIKE: Boolean = hasGlobalRegionScheduler()

    private fun hasGlobalRegionScheduler(): Boolean =
        try {
            Bukkit::class.java.getMethod("getGlobalRegionScheduler")
            true
        } catch (_: NoSuchMethodException) {
            false
        }

    /**
     * Returns whether Paper/Folia region schedulers are available.
     *
     * @return `true` on Paper and Folia; `false` on Spigot/Bukkit
     */
    @JvmStatic
    fun isFoliaLike(): Boolean = FOLIA_LIKE

    /**
     * Runs a task on the global region (Paper/Folia) or the main server thread (Spigot/Bukkit).
     *
     * @param plugin owning plugin
     * @param task work to execute
     */
    @JvmStatic
    fun runGlobal(
        plugin: Plugin,
        task: Runnable,
    ) {
        if (FOLIA_LIKE) {
            Bukkit.getGlobalRegionScheduler().execute(plugin, task)
        } else {
            Bukkit.getScheduler().runTask(plugin, task)
        }
    }

    /**
     * Schedules a repeating global-region (or main-thread) timer.
     *
     * @param plugin owning plugin
     * @param task work to execute each period
     * @param delayTicks ticks before the first run
     * @param periodTicks ticks between runs
     * @return opaque task handle suitable for [cancel]
     */
    @JvmStatic
    fun runGlobalTimer(
        plugin: Plugin,
        task: Runnable,
        delayTicks: Long,
        periodTicks: Long,
    ): Any =
        if (FOLIA_LIKE) {
            Bukkit.getGlobalRegionScheduler().runAtFixedRate(
                plugin,
                { task.run() },
                maxOf(1L, delayTicks),
                maxOf(1L, periodTicks),
            )
        } else {
            Bukkit.getScheduler().runTaskTimer(plugin, task, delayTicks, periodTicks)
        }

    /**
     * Runs a task asynchronously for I/O or other off-thread work.
     *
     * @param plugin owning plugin
     * @param task work to execute off the server thread
     */
    @JvmStatic
    fun runAsync(
        plugin: Plugin,
        task: Runnable,
    ) {
        if (FOLIA_LIKE) {
            Bukkit.getAsyncScheduler().runNow(plugin) { task.run() }
        } else {
            Bukkit.getScheduler().runTaskAsynchronously(plugin, task)
        }
    }

    /**
     * Runs a task on the player's owning region (Folia/Paper) or the main thread (Spigot/Bukkit).
     *
     * Use this when messaging or touching a player other than the current command sender so Folia
     * region affinity is respected.
     *
     * @param plugin owning plugin
     * @param player entity whose region owns the task
     * @param task work to execute
     */
    @JvmStatic
    fun runAtEntity(
        plugin: Plugin,
        player: Player,
        task: Runnable,
    ) {
        if (FOLIA_LIKE) {
            player.scheduler.run(plugin, { task.run() }, null)
        } else {
            Bukkit.getScheduler().runTask(plugin, task)
        }
    }

    /**
     * Runs a delayed task on the player's owning region (Folia/Paper) or the main thread
     * (Spigot/Bukkit).
     *
     * @param plugin owning plugin
     * @param player entity whose region owns the task
     * @param task work to execute
     * @param delayTicks ticks to wait before running
     */
    @JvmStatic
    fun runAtEntityLater(
        plugin: Plugin,
        player: Player,
        task: Runnable,
        delayTicks: Long,
    ) {
        val delay = maxOf(1L, delayTicks)
        if (FOLIA_LIKE) {
            player.scheduler.runDelayed(plugin, { task.run() }, null, delay)
        } else {
            Bukkit.getScheduler().runTaskLater(plugin, task, delay)
        }
    }

    /**
     * Cancels a task returned by [runGlobalTimer].
     *
     * @param task opaque Folia or Bukkit task handle; ignored when `null`
     */
    @JvmStatic
    fun cancel(task: Any?) {
        if (task == null) {
            return
        }
        try {
            task.javaClass.getMethod("cancel").invoke(task)
            return
        } catch (_: Exception) {
            // Fall through to BukkitTask handling.
        }
        if (task is BukkitTask) {
            task.cancel()
        }
    }
}
